cfme-performance/tests/workloads/test_provisioning.py

```diff
@@ -0,0 +1,99 @@
+"""Runs Provisioning Workload."""
+from utils.appliance import clean_appliance
+from utils.appliance import get_server_roles_workload_provisioning
+from utils.appliance import set_server_roles_workload_provisioning
+from utils.appliance import wait_for_miq_server_ready
+from utils.grafana import get_scenario_dashboard_url
+from utils.log import logger
+from utils.providers import add_providers
+from utils.providers import retire_provisioned_vms
+from utils.providers import provision_vm
+from utils.providers import get_template_guids
+from utils.smem_memory_monitor import SmemMemoryMonitor
+from utils.ssh import SSHClient
+from utils.workloads import get_provisioning_scenarios
+from itertools import cycle
+import time
+import pytest
+
+
+@pytest.mark.parametrize('scenario', get_provisioning_scenarios())
+def test_provisioning(request, scenario):
+    """Runs through provisioning scenarios using the REST API to
+    continously order a catalog item(s) and provision a VM for a specified period of time.
+    Memory Monitor creates graphs and summary at the end of each scenario."""
+
+    from_ts = int(time.time() * 1000)
+    ssh_client = SSHClient()
+    logger.debug('Scenario: {}'.format(scenario['name']))
+
+    clean_appliance(ssh_client)
+
+    monitor_thread = SmemMemoryMonitor(SSHClient(), 'workload-provisioning', scenario['name'],
+        'Provisioning', get_server_roles_workload_provisioning(separator=', '),
+        ', '.join(scenario['providers']))
+
+    def cleanup_workload(scenario, from_ts):
+        starttime = time.time()
+        to_ts = int(starttime * 1000)
+        g_url = get_scenario_dashboard_url(scenario, from_ts, to_ts)
+        logger.debug('Started cleaning up monitoring thread.')
+        monitor_thread.grafana_url = g_url
+        monitor_thread.signal = False
+        monitor_thread.join()
+        retire_provisioned_vms(scenario['vm_names'])
+        timediff = time.time() - starttime
+        logger.info('Finished cleaning up monitoring thread in {}'.format(timediff))
+    request.addfinalizer(lambda: cleanup_workload(scenario, from_ts))
+
+    monitor_thread.start()
+
+    wait_for_miq_server_ready(poll_interval=2)
+    set_server_roles_workload_provisioning(ssh_client)
+    add_providers(scenario['providers'])
+    logger.info('Sleeping for Refresh: {}s'.format(scenario['refresh_sleep_time']))
+    time.sleep(scenario['refresh_sleep_time'])
+
+    guid_list = get_template_guids(scenario['templates'])
+    guid_cycle = cycle(guid_list)
+    providers_num = len(scenario['providers'])
+    cleanup_size = scenario['cleanup_size']
+    total_time = scenario['total_time']
+    time_between_provision = scenario['time_between_provision']
+    total_provisioned_vms = 0
+    provisioned_vms = 0
+    provision_order = []
```

Comment:
This should be defined above the finalizer and passed into the finalizer.  Then the finalizer can pop the remaining vms off the list and attempt to retire them.